### DISCUSSION TITLE
Yet another question about the use cases for duckdb

### DISCUSSION THREAD

**Discussion by iliya-malecki** [duckdb/duckdb#10856] [Q&A]
I have googled extensively and tried my best, but I still struggle to understand when to consider using duckdb, sorry! Can you provide an example with comparisons?

I mean, if I need to do data wrangling, my default is polars. As the dataset size grows, I still try to stick with polars for as long as possible, before moving to spark - because something something hand wave hand wave distributed system overhead.

If my use case is more about databases, I'd use postgresql. If it is also write-heavy, I'd still use postgresql. If I need to do it at scale, and I don't have complex reads, I look into document-based. If I do need complex reads as well as frequent writes at scale, I start ~~crying~~ looking into fancy systems like Cassandra but that is way outside of the potential duckdb spot.

So the question is: what's the vacant spot? What tech does duckdb replace? Is it specifically polars? But then I don't have python and ".apply" so I can't even query openai while wrangling, right? I know it's not just a less flexible version of polars, because then it wouldn't get so much attention, so please help! Right now I only use duckdb for non-equi joins :)

**Comment by szarnyasg** [comment]
Hello @iliya-malecki, DuckDB is geared towards read-heavy, analytical workloads, so regarding its target workload, it's quite similar to Polars.

The way most users interact with DuckDB (loading/modifying the data, configuring the system, running queries) is SQL. DuckDB's SQL dialect is very close to PostgreSQL's one. This allows users not only to leverage their existing SQL knowledge but also to integrate with a tools that support the PostgreSQL dialect.

DuckDB made a lot of improvements in its [out-of-core capabilities](https://duckdb.org/docs/guides/performance/how_to_tune_workloads#larger-than-memory-workloads-out-of-core-processing). Therefore, if you have a fast disk (NVMe SSD), you can spill on it and you  can likely still finish larger-than-memory workloads without the need to start a Spark cluster.

DuckDB also has a lot of database features: an on-disk persistent format that can be instantly reloaded between session, a sophisticated SQL query optimizer, transactions that can be rolled back on abort, integrations with other DBMSs (Postgres / MySQL / SQLite, etc.).

DuckDB can [read Polars dataframes](https://duckdb.org/docs/guides/python/polars), so you can try to run a few queries of your workload in DuckDB and see whether it's a good fit.

> But then I don't have python and ".apply" so I can't even query openai while wrangling, right?

I'm sorry – I don't understand this. You can use DuckDB from Python, and you can return the results as (Pandas) dataframes. ChatGPT works reasonably well for SQL queries.

HTH

**Reply by iliya-malecki** [reply]
That's curious! 
Regarding my statement about openai, what I meant was that running data wrangling in python has the benefit of the proverbial ".apply", meaning I can apply an arbitrary python function to each element of a column. This means i can leverage stuff python is handy for, if I'm willing to tank the performance - for example, i can embed some text into a vector space with a rest call to openai

**Reply by szarnyasg** [reply]
DuckDB also allows registering and using user-defined functions in Python, see https://duckdb.org/docs/api/python/function
We have a blog post with more examples: https://duckdb.org/2023/07/07/python-udf

**Reply by iliya-malecki** [reply]
Wow okay that is really useful, along with strong out-of-core I can see how this is the swiss knife of the ds workflow
